src/core/schema.service.impl.ts

```diff
@@ -70,36 +70,57 @@ const deleteFromArray = (key: string) => (data: object) => (valueToDelete: objec
 const getArray = (key: string) => (data: object) => {
   return data[key];
 };
-const addReference = (schema: JsonSchema, variable: string, pathToContainment: string) =>
+const addReference = (schema: JsonSchema, identifyingProperty: string, variable: string) =>
   (root: Object, data: Object, toAdd: object) => {
-    const containment = pathToContainment
-      .split('/')
-      .reduce(
-        (elem, path) => {
-          if (path === '#' || path === '') {
-            return elem;
-          }
 
-          return elem[path];
-        },
-        root
-      );
-    const index = (containment as Object[]).indexOf(toAdd);
+    const refValue = toAdd[identifyingProperty];
     if (schema.properties[variable].type === 'array') {
       if (!data[variable]) {
         data[variable] = [];
       }
-      data[variable].push(index);
+      data[variable].push(refValue);
     } else {
-      data[variable] = index;
+      data[variable] = refValue;
     }
   };
-const getReference = (href: string, variable: string, variableWrapped: string) =>
-  (root: Object, data: Object) => {
-    const variableValue = data[variable];
-    const pathToObject = href.replace(variableWrapped, variableValue);
+const resolveRef = (schema: JsonSchema, findTargets: (rootData: Object) => Object[],
+                    identifyingProperty: string, variable: string) =>
+  (rootData: Object, data: Object) => {
+    if (_.isEmpty(data) || _.isEmpty(identifyingProperty)) {
+      return null;
+    }
+    // get all objects that could be referenced.
+    const candidates = findTargets(rootData);
+
+    if (schema.properties[variable].type === 'array') {
```

Comment:
I think the `type` keyword has to stay: The test cases fail on removal because the whole schema for the variable is compared to `array` which will always fail.